### DISCUSSION TITLE
getInitialProps in custom _document never gets called server side

### DISCUSSION THREAD

**Discussion by dagda1** [vercel/next.js#33022] [Help]
I have the following custom `_document` with a `getInitialProps` function that never gets called on the server.

Do I need to configure this somehow in `next.config`?

```ts
class FrontendDXDocument extends Document<Props> {
  static async getInitialProps(ctx: DocumentContext): Promise<Props> {
    const docProps = await ctx.defaultGetInitialProps(ctx);

    if (!ctx.res) {
      return docProps;
    }

    // NEVER GETS HERE

    const nonce = nanoid();

    let contentSecurityPolicy = '';
    contentSecurityPolicy = createContentSecurityPolicy(nonce); // `default-src 'self'; style-src 'unsafe-inline'; script-src 'self' 'unsafe-eval';`;

    ctx.res.setHeader('Content-Security-Policy', contentSecurityPolicy);
    return { ...docProps, nonce };
  }
```

**Comment by leerob** [comment]
Hey folks, wanted to swing back here with an update. After digging through many different issues and discussions, I've [**made a new page in the documentation**](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy) ([PR](https://github.com/vercel/next.js/pull/54601)) specifically for Content Security Policy and nonces. This docs page: 

- Explains how to generate a `nonce` with Middleware
- Shows how to consume the `nonce` in a route with `headers()`
- Shows a complete CSP without needing to use any `unsafe`
- Shows how to ignore the `nonce` Middleware from running on prefetches / static assets

Further, we've patched some bugs and made improvements to `nonce` handling in Next.js that will be available in the latest `canary` version (for those of you time traveling from the future, upgrade to Next.js 13.5). We also updated the `with-strict-csp` example in the `examples/` folder, which is backlinked from the [new documentation page](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy).

Really hope this helps out, thank you all 🙏 I'll be closing this discussion out. To continue the discussion, [please go here](https://github.com/vercel/next.js/discussions/54907).
